Snowflake.Data/Core/Session/CreateSessionTokens.cs

```diff
@@ -0,0 +1,43 @@
+using System;
+using System.Collections.Generic;
+using Snowflake.Data.Core.Session;
+
+namespace Snowflake.Data.Core.Session
+{
+    internal class CreateSessionTokens: ICreateSessionTokens
+    {
+        internal long _timeout { get; set; }  = Timeout;
+        private const long Timeout = 30000; // 30 seconds as default
+        private readonly object _tokenLock = new object();
+        private readonly List<CreateSessionToken> _tokens = new List<CreateSessionToken>();
+        private int _tokenCount = 0;
+        
+        public CreateSessionToken BeginCreate()
+        {
+            lock (_tokenLock)
+            {
+                var token = new CreateSessionToken(_timeout);
+                _tokens.Add(token);
+                var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+                _tokens.RemoveAll(t => t.IsExpired(now));
+                _tokenCount = _tokens.Count;
+                return token;
+            }
+        }
+
+        public void EndCreate(CreateSessionToken token)
+        {
+            lock (_tokenLock)
+            {
+                var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+                _tokens.RemoveAll(t => token.Id == t.Id || t.IsExpired(now));
```

Comment:
RemoveAll will remove the handle from _tokens. If the session creation would succeed after given timeout, then a session would be created and we would may exceed max pool size. So the assumption is that the timeout would be reasonable to avoid such cases from one hand but from the other hand even if we allow in some scenario an additional session is it a real problem?

Reply:
It depends, what if this timeout is badly set by a user

Reply:
we agreed to leave as is for now